Make more objects spray paintable (Reviving #31328)

whatston3 opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->

Reviving #31328 with most suggestions added in, and the branch itself brought to a working state.  Many thanks to Ertanic for the previous work.

Adds the ability for the spray painter to recolour gas canisters, lockers (standing and wall lockers), crates, and paint decals on the floor.  The spray painter now consumes charges from compressed dye canisters (15 per canister at time of writing), and the rate of consumption depends on the item being printed (e.g. pipes, 1; lockers, 3) - this is similar to compressed matter used by an RCD.

Painted items take on the appearance of one of a set of selectable objects of the same type (crates look like crates, canisters look like other canisters), and can be inspected to see that they've been painted until the paint dries (currently takes 15 minutes).  All painted decals are uncleanable (it is assumed that adding/removing tiles is sufficient to remove the decals - if not, the painter itself could remove them with another decal placement mode in the future).

Spray painting still works with doafters, except for decal painting, which happens instantly, more similar to a crayon.  Unlike the crayon, the spray painter's decal painting can be toggled using an alt-verb to prevent unwanted decal placement.

Spray painter ammo is available in the YouTool beside the spray painter, and additional ammo can be printed from the autolathe for 150 steel and 50 plastic.  The existing spray painter recipe in the autolathe now prints with no charges.

TODO:
- [x] Make decal selection UI respect the state the painter was left in
- [x] Add the ability to place decals without snapping to grid
- [x] Review secure vs. non-secure paintables (although the sprites will look a bit odd with an empty keyhole/light, it's not too farfetched)
- [x] Consider subtabs for subcategories (e.g. Crates->Steel/Plastic/Secure)
- [x] Add a way to create compressed dye
- [x] More meticulous review & cleanup

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->

Nice quality of life, especially for paintable gas canisters.  It's nice to be able to repaint things to decorate a room, or to easily identify which gases are which without needing a bunch of oddball canisters.  Might allow for some amount of subterfuge by repainting lockers.

## Technical details
<!-- Summary of code changes for easier review. -->

There are three new components, PaintableComponent for items that _can be_ painted by a spray painter (replacing PaintableAirlockComponent) - this component associates an object with a particular _paintable group_, PaintedComponent for items that _have been_ painted, and SprayPainterAmmo for items that can recharge the spray painter.

Two new prototypes have been added to organize paintable items.  The main prototype used is the PaintableGroupPrototype, which maps _paintable styles_ (arbitrary strings) to entity prototypes and keeps info on how long and costly it is to paint an object in this group.  The other prototype, PaintableGroupCategoryPrototype, collects paintable groups in lists for display in the spray painter UI.

New tabs have been added to the spray painter's UI, and the BUI has been swapped over to use an Update call and the state of the component itself, using the GasPressurePump as a template.  If the component state is too bulky for this, and would benefit from using delta states, this can be fixed, but I have not done this.

The door system's BaseRSI visuals has been removed in favour of PaintableVisuals, which send off the prototype to be used as a basis for its appearance.  What exactly that means is left to the individual systems - a gas canister works differently than an airlock, which works differently than a crate.

Additionally, upon being painted, a new event is sent out for items that need special handling (e.g. gas canisters updating the entity they will turn into upon breaking).  Copying the destructible behaviour for a gas canister is odd, but the canister breaking and magically changing colour is also odd - I kept this around since it provided better visual consistency.

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

The video below shows the spray painter in use.  The amount of charges in each ammo entity has been kept at 15.  Note that each clicked entity that does not paint (e.g. freezers, mining/external airlocks, trash bins) has been explicitly blacklisted here (via a null paintable group).  Each tab represents a paintable category, and each subtab represents a paintable group within that category (e.g. wallmount closets)

https://github.com/user-attachments/assets/f25c429d-4c12-4509-8645-f053848a8be1

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

DoorVisuals.BaseRSI has been removed - any instance should be replaced with PaintableVisuals.Prototype.  PaintableAirlockComponent has been removed - use PaintableComponent instead.  Refer to existing airlocks as an example of use.  To exempt any airlocks/crates/etc. from being paintable, add a PaintableComponent with a null group.
AirlockGroupPrototype and AirlockDepartmentPrototype have been removed, replace their use with PaintableGroupPrototype and PaintableGroupCategoryPrototype.

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl: Ertanic and whatston3
- add: The spray painter can now paint gas canisters, crates, lockers, and floor decals.
- tweak: The spray painter now consumes compressed paint, available from the YouTool or printable in the autolathe.

Ertanic committed:
PaintableAirlockComponent and AirlockGroupPrototype have been replaced

Ertanic committed:
Slightly redesigned SprayPainterSystem for greater versatility

Ertanic committed:
Added handling of changes to the appearance of doors and storages

Ertanic committed:
PaintableGroup prototypes have been created

Ertanic committed:
Generating tabs with styles in the UI

Ertanic committed:
Fix error with undiscovered layer

Ertanic committed:
Merge branch 'master' into spray-painter-update

Ertanic committed:
Merge branch 'upstream' into spray-painter-update

Ertanic committed:
Slight improvement

Ertanic committed:
Removed unnecessary property

Ertanic committed:
The category for `PaintableGroup` was allocated to a separate prototype so that the engine itself would check if the category existed

Ertanic committed:
Merge branch 'upstream' into spray-painter-update

Ertanic committed:
Added canisters, but repainting doesn't work

Ertanic committed:
Added localization to styles

Ertanic committed:
Fix sprite changing

Ertanic committed:
Added the ability to paint canisters

Ertanic committed:
slight ui improvement

Ertanic committed:
Fix yamllinter errors

Ertanic committed:
Fix test

Ertanic committed:
The UI now remembers which tab was open

Ertanic committed:
Fix build (?)

Ertanic committed:
Rename

Ertanic committed:
Charges have been added to the spray painter

Ertanic committed:
Merge branch 'space-wizards:master' into spray-painter-update

Ertanic committed:
Added a charge texture for the spray painter

Ertanic committed:
Merge branch 'spray-painter-update' of https://github.com/Ertanic/space-station-14 into spray-painter-update

Ertanic committed:
Merge branch 'space-wizards:master' into spray-painter-update

Ertanic committed:
Now spray painter can paint decals

Ertanic committed:
Increased number of charges

Ertanic committed:
Spawning dummy objects has been replaced by PrototypeManager

Ertanic committed:
added a signature about the painting of the object

Ertanic committed:
Upstream

Ertanic committed:
fix

Ertanic committed:
Code commenting

Ertanic committed:
Merge branch 'master' into spray-painter-update

Ertanic committed:
Fix upstream

Ertanic committed:
Update Content.Shared/SprayPainter/Components/SprayPainterAmmo.cs

Co-authored-by: pathetic meowmeow <uhhadd@gmail.com>

Ertanic committed:
review

Ertanic committed:
Merge branch 'spray-painter-update' of https://github.com/Ertanic/space-station-14 into spray-painter-update

Ertanic committed:
Now decals can only be painted if the corresponding tab in the menu is open.

Ertanic committed:
Fixed a bug with pipe and decal tabs not being remembered

Ertanic committed:
Update EntityStorageVisualizerSystem.cs

Ertanic committed:
record

Ertanic committed:
Merge remote-tracking branch 'origin/spray-painter-update' into spray-painter-update

Ertanic committed:
loc

whatston3 committed:
Cleanup

whatston3 committed:
Revert electrified visuals

whatston3 committed:
more cleanup, fix charges, del ammo4

whatston3 committed:
no empty file, remove meta component

whatston3 committed:
closet exceptions, storage visualizer fixes

whatston3 committed:
enable/disable decal through alt-verb

whatston3 committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into 2025-05-10-the-cooler-spray-painter-update

whatston3 committed:
Fix missed merge conflicts

whatston3 committed:
fix snap offset, button event handlers

whatston3 committed:
simpler order, fix snap loc string

whatston3 committed:
Remove PaintableViz.BaseRSI, no decal item, A-Z

whatston3 committed:
State-respecting UI, BUI updates, FTL fixes

whatston3 committed:
revert DecalPlacerWindow changes

whatston3 committed:
revert unwanted changes, cleanup function order

whatston3 committed:
Limit SprayPainterAmmo write access to AmmoSystem

whatston3 committed:
Remove PaintedSystem

whatston3 committed:
spray paint ammo lathe recipe, youtool listing

whatston3 committed:
category as a list, groups as subtabs

whatston3 committed:
Restore inhand copyright in meta.json

whatston3 committed:
empty spray painter, recipe produces an empty one

whatston3 committed:
allow alpha on spray painter decals

whatston3 committed:
add comments

whatston3 commented:
Think this is about good to go.

Need to post an update to the description and show the thing working in practice, there've been considerable changes, but I largely think they make sense.  Biggest questionable changes, at least from my side:
- the tool has an alt-verb to enable/disable decal placement rather than relying on which tab was last open
- BUI now updates client-side based on received state updates (worked largely using the gas pressure pump as a template, though this one is much larger - might make sense to swap over to use deltas?)
- no reliance on indices in messages, everything's sent as a string
- groups can be selected separately as subtabs in the spray painter window
- every paintable group has a default value, as well as the decals and pipes.
- group association is stored in the category itself rather than by groups storing which category they belong to

whatston3 committed:
paintable wall lockers

whatston3 committed:
Restrict painting more objects

whatston3 commented:
To those that reviewed #31328 originally (@sowelipililimute, @MilonPL), the spray painter is generally functional here; I think this is ready for review.

slarticodefast commented on Content.Shared/SprayPainter/Components/SprayPainterComponent.cs:
```diff
@@ -29,9 +45,60 @@ public sealed partial class SprayPainterComponent : Component
+    /// <summary>
+    /// Whether or not the painter should be painting decals.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadOnly)]
```
any reason this one is not a datafield?

whatston3 replied:
No - mess on my part from when it was based on tabs.

whatston3 committed:
Suggested event changes, event cleanup

whatston3 commented on Content.Shared/SprayPainter/SprayPainterEvents.cs:
```diff
@@ -71,3 +102,33 @@ public SprayPainterPipeDoAfterEvent(Color color)
+/// <summary>
+/// An action raised on an item when it's spray painted.
+/// </summary>
+public sealed partial class EntityPaintedEvent : EntityEventArgs
```
Should be fixed with `6fa2a92`.

whatston3 committed:
component comments, fix ammo inhands

whatston3 committed:
uncleanable decals, dirty styles on mapinit

whatston3 committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into 2025-05-10-the-cooler-spray-painter-update

whatston3 commented:
Conflicts should be resolved, unsure why the bot hasn't removed the label.

Ertanic commented:
I think it would be a good idea to separate lockers with and without locks into different tabs, to avoid a visual bug, as is done with the crates. Including emergency wallmount lockers, it's pretty weird that the whole shape of the object changes, so I suggest making an “Emergency” tab.

whatston3 commented:
> I think it would be a good idea to separate lockers with and without locks into different tabs, to avoid a visual bug, as is done with the crates.

Unlocked versions of lockers tend to look fine.  Locked closets can look odd, so fair dues.

> Including emergency wallmount lockers, it's pretty weird that the whole shape of the object changes, so I suggest making an “Emergency” tab.

If the consensus is "give me a subtab for two whole things", then sure.

whatston3 committed:
organize paintables, separate emergency/closet grp

whatston3 commented:
Changes addressed.

Alphabetized all paintable groups, moved the prototypes out of the Entities folder into their own Paintables folder (pending a better name).

whatston3 committed:
fix categories newline at EOF

whatston3 committed:
airlock group whitespace cleanup

whatston3 committed:
realphabetize

iaada commented:
I don't think there needs to be a group for emergency wall closets. Yes they change shape, but that's really not breaking my immersion. Plenty of things change shape, even the spray painter already changes the shape of airlocks with the freezer, or the default window airlock.

whatston3 committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into 2025-05-10-the-cooler-spray-painter-update

whatston3 committed:
Clean up EntityStorageViz merge conflict markers

whatston3 commented on Content.Client/Atmos/EntitySystems/GasCanisterAppearanceSystem.cs:
```diff
@@ -0,0 +1,29 @@
+        if (!proto.TryGetComponent(out SpriteComponent? sprite, _componentFactory))
+            return;
+
+        old.LayerSetState(0, sprite.LayerGetState(0));
```
Here's a question for you - all of the SpriteSystem API seems to work with Entities.  Do I need to spawn a temporary entity from the prototype to get this to work, or is there something that would grab it from the prototype's components?

whatston3 committed:
Apply requested changes

sowelipililimute commented on Content.Client/SprayPainter/UI/SprayPainterGroup.xaml.cs:
```diff
@@ -0,0 +1,66 @@
+[GenerateTypedNameReferences]
+public sealed partial class SprayPainterGroup : BoxContainer
+{
+    public Action<SpriteListData>? OnButtonPressed;
```
```suggestion
    public event Action<SpriteListData>? OnButtonPressed;
```

sowelipililimute commented on Content.Client/SprayPainter/UI/SprayPainterWindow.xaml.cs:
```diff
@@ -15,13 +22,33 @@ public sealed partial class SprayPainterWindow : DefaultWindow
-    public Action<ItemList.ItemListSelectedEventArgs>? OnSpritePicked;
-    public Action<ItemList.ItemListSelectedEventArgs>? OnColorPicked;
+    // Events
+    public Action<string, string>? OnSpritePicked;
+    public Action<int, bool>? OnTabChanged;
+    public Action<ProtoId<DecalPrototype>>? OnDecalChanged;
+    public Action<ItemList.ItemListSelectedEventArgs>? OnSetPipeColor;
+    public Action<Color?>? OnDecalColorChanged;
+    public Action<int>? OnDecalAngleChanged;
+    public Action<bool>? OnDecalSnapChanged;
```
```suggestion
    public event Action<string, string>? OnSpritePicked;
    public event Action<int, bool>? OnTabChanged;
    public event Action<ProtoId<DecalPrototype>>? OnDecalChanged;
    public event Action<ItemList.ItemListSelectedEventArgs>? OnSetPipeColor;
    public event Action<Color?>? OnDecalColorChanged;
    public event Action<int>? OnDecalAngleChanged;
    public event Action<bool>? OnDecalSnapChanged;
```

sowelipililimute commented on Content.Client/Storage/Visualizers/EntityStorageVisualizerSystem.cs:
```diff
@@ -26,12 +31,34 @@ private void OnComponentInit(EntityUid uid, EntityStorageVisualsComponent comp,
+            || !AppearanceSystem.TryGetData<bool>(uid, StorageVisuals.Open, out var open, args.Component))
             return;
 
+        bool forceRedrawBase = false;
```
```suggestion
        var forceRedrawBase = false;
```

sowelipililimute commented on Content.Shared/SprayPainter/Components/PaintedComponent.cs:
```diff
@@ -0,0 +1,17 @@
+    /// <summary>
+    /// The time after which the entity is dried and does not appear as "freshly painted".
+    /// </summary>
+    [DataField, AutoNetworkedField, AutoPausedField]
```
```suggestion
    [DataField(customTypeSerializer: typeof(TimeOffsetSerializer)), AutoNetworkedField, AutoPausedField]
```

sowelipililimute reviewed (changes requested):
some minor nits, but otherwise looking good. thanks for picking this up!

whatston3 committed:
Apply suggestions from sowelipililimute's review

Co-authored-by: pathetic meowmeow <uhhadd@gmail.com>

whatston3 commented:
> some minor nits, but otherwise looking good. thanks for picking this up!

Thanks for the review.  Suggestions reviewed - everything looked good, changes applied.

whatston3 committed:
betrayal most foul

Centronias commented on Content.Shared/SprayPainter/SprayPainterEvents.cs:
```diff
@@ -71,3 +91,19 @@ public SprayPainterPipeDoAfterEvent(Color color)
     public override DoAfterEvent Clone() => this;
 }
+
+/// <summary>
+/// An action raised on an item when it was spray painted.
+/// </summary>
+[ByRefEvent]
+public partial record struct EntityPaintedEvent(EntityUid? user, EntityUid tool, EntProtoId prototype, ProtoId<PaintableGroupPrototype> group)
+{
+    /// <summary>The entity painting this item.</summary>
+    public EntityUid? User = user;
+    /// <summary>The entity used to paint this item.</summary>
+    public EntityUid Tool = tool;
+    /// <summary>The prototype used to generate the new painted appearance.</summary>
+    public EntProtoId Prototype = prototype;
+    /// <summary>The group of item being painted (e.g. airlocks with glass, canisters).</summary>
+    public ProtoId<PaintableGroupPrototype> Group = group;
+}
```
```suggestion
/// <summary>
/// An action raised on an item when it was spray painted.
/// </summary>
/// <param name="User">The entity painting this item.</param>
/// <param name="Tool">The entity used to paint this item.</param>
/// <param name="Prototype">The prototype used to generate the new painted appearance.</param>
/// <param name="Group">The group of item being painted (e.g. airlocks with glass, canisters).</param>
[ByRefEvent]
public partial record struct EntityPaintedEvent(
    EntityUid? User,
    EntityUid Tool,
    EntProtoId Prototype,
    ProtoId<PaintableGroupPrototype> Group);
```

Record structs don't need member fields.

whatston3 replied:
Fair dues.  Changes applied.

whatston3 committed:
Remove members from EntityPaintedEvent

sowelipililimute reviewed (approved):
this is looking good on my end, thanks for picking this PR up

whatston3 committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into 2025-05-10-the-cooler-spray-painter-update

whatston3 committed:
No emerg. group, steelsec to secure, locker/closet

whatston3 committed:
Enable repainting the medical wall locker

whatston3 committed:
comments, no flags on PaintableVisuals

whatston3 commented:
Added a wall locker category vs. wall closets (given that now there are two of them), collapsed the emergency lockers into the wall closets, added the evac repair locker and wall locker, alphabetized the FTL, restored a few comments, removed the Flags attribute from PaintedVisuals (only one value, not used as a set of flags any more, meaningless).

whatston3 committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into 2025-05-10-the-cooler-spray-painter-update

TiniestShark commented:
So overall I'm going to give this a thumbs up, this is something I've been wanting personally for a long time if only to make organization way easier. Tested it out and everything seems to work pretty all right (Giving them an ammo is a smart choice as well so you don't have a tider changing every door/locker on station).

My only gripe is the unlocked lockers category. Being able to spray paint generic lockers as secured lockers but without the light is strange and I'd rather just have those removed. You can already build secure closets/lockers fairly easily so I don't particularly see the benefit of coloring generics with an ill-fitting palette. I get the desire but it just looks incorrect. Same goes for the secure wall lockers/closets, I'd rather those options just be kept to their specific type.

Being able to remove decals would be nice on tiles that you can't pry up like plating and such (or at least not pry safely anyway), but I suppose you can also just cover them. It's fine.

whatston3 commented:
> My only gripe is the unlocked lockers category. Being able to spray paint generic lockers as secured lockers but without the light is strange and I'd rather just have those removed. You can already build secure closets/lockers fairly easily so I don't particularly see the benefit of coloring generics with an ill-fitting palette. I get the desire but it just looks incorrect. Same goes for the secure wall lockers/closets, I'd rather those options just be kept to their specific type.

Alright.

> Being able to remove decals would be nice on tiles that you can't pry up like plating and such (or at least not pry safely anyway), but I suppose you can also just cover them. It's fine.

Bigger ask.  How should it work?  A third mode?  Should it follow the same snap logic?  If not snapping, where does it go?

whatston3 committed:
Remove locked variants from closets/wall closets

whatston3 committed:
removable decals

whatston3 committed:
off value consistency

whatston3 committed:
can't paint away those bones

whatston3 committed:
fix precedence

whatston3 commented:
> Being able to remove decals would be nice on tiles that you can't pry up like plating and such (or at least not pry safely anyway), but I suppose you can also just cover them. It's fine.

@TiniestShark pushed an update for this.  Video included below.

There is now a third paint mode that's selectable with alt-Z or right-click (all cycle through Off, Add, Remove).  Dirt is no longer spray paintable (add or remove).  In the video, interactions with the painter set to off are shown to do nothing, then new decals are spray painted with snap both on and off, then decals are removed with snap off and on.  Finally, a standing locker and closet are painted, showing the locked styles no longer show up for closets.

https://github.com/user-attachments/assets/0d771de0-9bba-4293-a54c-d00a985a4563

whatston3 commented:
> I was going to suggest player painted decals be cleanable with Space cleaner

Figured it was better for janitor/painter sanity to know that decals won't be bulldozed by errant spray.

whatston3 committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into 2025-05-10-the-cooler-spray-painter-update

TiniestShark commented:
> Figured it was better for janitor/painter sanity to know that decals won't be bulldozed by errant spray.

Yeah, crayon artists are already a high cause of death to janitors not paying attention so it's definitely for the best.

Velken commented:
I'd love to see disposal pipes also paintable:
-allows disposal to be vidually be different from transport/mail pipes
-lets transport/mail be colored by department, making it less of a headache to engineering fixing reroutes/damages (currently you kinda have to memorize)
-lets antags make less suspicious pipe reroutes by also painting them
-helps when mapping so you don't have to follow the whole pipe to remember where it goes

whatston3 commented:
> I'd love to see disposal pipes also paintable: -allows disposal to be vidually be different from transport/mail pipes -lets transport/mail be colored by department, making it less of a headache to engineering fixing reroutes/damages (currently you kinda have to memorize) -lets antags make less suspicious pipe reroutes by also painting them -helps when mapping so you don't have to follow the whole pipe to remember where it goes

Fair ask, but I'd prefer not to expand the scope for that.  This doesn't store color information by key - it could, and then you could have a list of colourable object types selectable in the UI.  That said, I don't think that distinguishing disposals pipes by network is nearly as useful as with gas pipes.

whatston3 committed:
Remove AirlockDepartment, AirlockGroup protos

Both unused.

whatston3 committed:
whitelist consistency re: ammo component

whatston3 commented:
Minor cleanup pass done, removed a few unused prototypes (documented in the breaking changes section of the PR description), fixed a discrepancy between the CE belt and utility belt whitelists.

whatston3 committed:
add standing emergency closet styles

whatston3 committed:
alphabetize the spray painter listings

whatston3 committed:
Merge branch 'master' of https://github.com/space-wizards/space-station-14 into 2025-05-10-the-cooler-spray-painter-update

slarticodefast commented on Content.Shared/SprayPainter/Components/SprayPainterComponent.cs:
```diff
@@ -29,9 +45,82 @@ public sealed partial class SprayPainterComponent : Component
+/// <summary>
+/// A set of operating modes for decal painting.
+/// </summary>
+public enum DecalPaintMode : byte
```
This needs `[Serializable, NetSerializable]`
